1_The_Physical/Level_2_Stateful_Guess_Point/Shared/Student/src/domain/student/core/student.ts

```diff
@@ -0,0 +1,137 @@
+import { v4 as uuid } from "uuid";
+
+import {
+  Email,
+  EmailValidationError,
+  FirstName,
+  FirstNameValidationError,
+  LastName,
+  LastNameValidationError,
+} from "../value-objects";
+import { Result } from "../../../shared/result";
+import { AggregateRoot } from "../../core/aggregate-root";
+import { DomainEvent } from "../../core/domain-event";
+import { StudentCreated } from "../events/student-created";
+import { FirstNameUpdated } from "../events/first-name-updated";
+import { LastNameUpdated } from "../events/last-name-updated";
+
+interface StudentInputProps {
+  firstName: string;
+  lastName: string;
+}
+
+interface StudentState {
+  firstName: FirstName;
+  lastName: LastName;
+  email: Email;
+}
+
+interface StudentValidationError {
+  firstName?: FirstNameValidationError;
+  lastName?: LastNameValidationError;
+  email?: EmailValidationError;
+}
+
+export class Student implements AggregateRoot<StudentState> {
+  public readonly id: string;
+  public readonly state: StudentState;
+  public readonly events: DomainEvent[] = [];
+
+  private constructor(props: StudentState) {
+    this.id = uuid();
+    this.state = props;
+    this.events.push(new StudentCreated(this));
+  }
+
+  public static create(
+    props: StudentInputProps
+  ): Result<Student, StudentValidationError> {
+    const firstNameResult = FirstName.create(props.firstName);
+    const lastNameResult = LastName.create(props.lastName);
+    const emailResult = Result.combine(firstNameResult, lastNameResult).flatMap(
+      ([FirstName, LastName]) =>
+        Student.generateEmail(FirstName.value, LastName.value)
+    );
+
+    const errors: StudentValidationError = {};
+    if (firstNameResult.isFailure()) {
+      errors.firstName = firstNameResult.error;
+    }
+    if (lastNameResult.isFailure()) {
+      errors.lastName = lastNameResult.error;
+    }
+    if (emailResult.isFailure() && !(emailResult.error instanceof Array)) {
```

Comment:
Regarding

```ts
!(emailResult.error instanceof Array)
```

It looks like an abstraction is not expressed properly or there is a leaky abstraction. What if it's an array? I don't see what handles this case. If it's an internal illegal state of `Result`, why then `Student` needs to care about it?

Reply:
Indeed, I had rushed through this part a bit. This has been resolved with the refactoring of the `Result` class.